File: Backend/models/modelLogModel.js
const db = require('../config/database');

const ModelLog = {
  findAll: async () => {
    const [rows] = await db.query('SELECT * FROM model_logs ORDER BY created_at DESC');
    return rows;
  },
  findBest: async () => {
    const [rows] = await db.query('SELECT * FROM model_logs WHERE status = "completed" ORDER BY rmse ASC LIMIT 1');
    return rows[0];
  },
  findLatest: async () => {
    const [rows] = await db.query('SELECT * FROM model_logs ORDER BY created_at DESC LIMIT 1');
    return rows[0];
  },
  findById: async (id) => {
    const [rows] = await db.query('SELECT * FROM model_logs WHERE id = ?', [id]);
    return rows[0];
  },
  create: async (data) => {
    const [result] = await db.query(
      'INSERT INTO model_logs (units, learning_rate, dropout_rate, batch_size, epochs, rmse, status, completed_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?)',
      [data.units, data.learning_rate, data.dropout_rate, data.batch_size, data.epochs, data.rmse || null, data.status || 'pending', data.completed_at || null]
    );
    return result.insertId;
  },
  update: async (id, data) => {
    const fields = [];
    const values = [];
    if (data.units !== undefined) { fields.push('units = ?'); values.push(data.units); }
    if (data.learning_rate !== undefined) { fields.push('learning_rate = ?'); values.push(data.learning_rate); }
    if (data.dropout_rate !== undefined) { fields.push('dropout_rate = ?'); values.push(data.dropout_rate); }
    if (data.batch_size !== undefined) { fields.push('batch_size = ?'); values.push(data.batch_size); }
    if (data.epochs !== undefined) { fields.push('epochs = ?'); values.push(data.epochs); }
    if (data.rmse !== undefined) { fields.push('rmse = ?'); values.push(data.rmse); }
    if (data.status !== undefined) { fields.push('status = ?'); values.push(data.status); }
    if (data.completed_at !== undefined) { fields.push('completed_at = ?'); values.push(data.completed_at); }
    if (fields.length === 0) return 0;
    values.push(id);
    const [result] = await db.query(`UPDATE model_logs SET ${fields.join(', ')} WHERE id = ?`, values);
    return result.affectedRows;
  }
};

module.exports = ModelLog;

File: backend_python/models/model_log_model.py
from config.database import execute, query, query_one


class ModelLog:
    @staticmethod
    def find_all():
        return query("SELECT * FROM model_logs ORDER BY created_at DESC")

    @staticmethod
    def find_best():
        return query_one(
            'SELECT * FROM model_logs WHERE status = "completed" ORDER BY rmse ASC LIMIT 1'
        )

    @staticmethod
    def find_latest():
        return query_one("SELECT * FROM model_logs ORDER BY created_at DESC LIMIT 1")

    @staticmethod
    def find_by_id(row_id):
        return query_one("SELECT * FROM model_logs WHERE id = %s", (row_id,))

    @staticmethod
    def create(data):
        _, insert_id = execute(
            "INSERT INTO model_logs (units, learning_rate, dropout_rate, batch_size, epochs, rmse, status, completed_at) "
            "VALUES (%s, %s, %s, %s, %s, %s, %s, %s)",
            (
                data["units"],
                data["learning_rate"],
                data["dropout_rate"],
                data["batch_size"],
                data["epochs"],
                data.get("rmse"),
                data.get("status") or "pending",
                data.get("completed_at"),
            ),
        )
        return insert_id

    @staticmethod
    def update(row_id, data):
        fields = []
        values = []
        for key in (
            "units",
            "learning_rate",
            "dropout_rate",
            "batch_size",
            "epochs",
            "rmse",
            "status",
            "completed_at",
        ):
            if key in data:
                fields.append(f"{key} = %s")
                values.append(data[key])
        if not fields:
            return 0
        values.append(row_id)
        rowcount, _ = execute(
            f"UPDATE model_logs SET {', '.join(fields)} WHERE id = %s", tuple(values)
        )
        return rowcount


class Comparison:
    @staticmethod
    def find_by_type(model_type):
        return query_one("SELECT * FROM comparison_metrics WHERE model_type = %s", (model_type,))

    @staticmethod
    def upsert(model_type, data):
        execute(
            "INSERT INTO comparison_metrics (model_type, mape, rmse, mae, r2) VALUES (%s, %s, %s, %s, %s) "
            "ON DUPLICATE KEY UPDATE mape = VALUES(mape), rmse = VALUES(rmse), mae = VALUES(mae), r2 = VALUES(r2)",
            (model_type, data["mape"], data["rmse"], data["mae"], data["r2"]),
        )
